runtime/js/40_testing.js

```diff
@@ -75,56 +150,67 @@
       // We're checking diff because one might spawn HTTP server in the background
       // that will be a pending async op before test starts.
       const dispatchedDiff = post.opsDispatchedAsync - pre.opsDispatchedAsync;
       const completedDiff = post.opsCompletedAsync - pre.opsCompletedAsync;
 
+      if (dispatchedDiff === completedDiff) return;
+
       const details = [];
       for (const key in post.ops) {
-        const dispatchedDiff = Number(
-          post.ops[key]?.opsDispatchedAsync -
-            (pre.ops[key]?.opsDispatchedAsync ?? 0),
-        );
-        const completedDiff = Number(
-          post.ops[key]?.opsCompletedAsync -
-            (pre.ops[key]?.opsCompletedAsync ?? 0),
-        );
-
-        if (dispatchedDiff !== completedDiff) {
-          details.push(`
-  ${key}:
-    Before:
-      - dispatched: ${pre.ops[key]?.opsDispatchedAsync ?? 0}
-      - completed: ${pre.ops[key]?.opsCompletedAsync ?? 0}
-    After:
-      - dispatched: ${post.ops[key].opsDispatchedAsync}
-      - completed: ${post.ops[key].opsCompletedAsync}`);
+        const preOp = pre.ops[key] ??
+          { opsDispatchedAsync: 0, opsCompletedAsync: 0 };
+        const postOp = post.ops[key];
+        const dispatchedDiff = postOp.opsDispatchedAsync -
+          preOp.opsDispatchedAsync;
+        const completedDiff = postOp.opsCompletedAsync -
+          preOp.opsCompletedAsync;
+
+        if (dispatchedDiff > completedDiff) {
+          const [name, hint] = OP_DETAILS[key];
+          const count = dispatchedDiff - completedDiff;
+          let message = `${count} async operation${
+            count === 1 ? "" : "s"
+          } to ${name} ${
+            count === 1 ? "was" : "were"
+          } started in this test, but never completed. This is often caused by not ${hint}.`;
+          const traces = [];
+          for (const [id, { opName, stack }] of postTraces) {
+            if (opName !== key) continue;
+            if (MapPrototypeHas(preTraces, id)) continue;
+            ArrayPrototypePush(traces, stack);
+          }
+          if (traces.length === 1) {
+            message += " The operation was started here:\n";
+            message += traces[0];
+          } else if (traces.length > 1) {
+            message += " The operations were started here:\n";
+            message += StringPrototypeJoin(traces, "\n\n");
+          }
+          ArrayPrototypePush(details, message);
+        } else if (dispatchedDiff < completedDiff) {
+          const [name] = OP_DETAILS[key];
+          const count = completedDiff - dispatchedDiff;
+          ArrayPrototypePush(
+            details,
+            `${count} async operation${count === 1 ? "" : "s"} to ${name} ${
+              count === 1 ? "was" : "were"
+            } started before this test, but ${
+              count === 1 ? "was" : "were"
+            } completed during the test. Async operations should not complete in a test if they were not started in that test.`,
+          );
```

Comment:
This is better, but I think still not exactly correct if I'm understanding the code correctly. I'm not sure we should be counting all ops, but instead only counting ops that weren't started when the test ran.

Example:

1. User starts Test A without sanitizers. It finishes with 1 pending op and Test B starts.
2. While Test B is running, the op from Test A completes.
3. Test B completes with 1 pending op.
4. Sanitizers incorrectly pass because the diff is 0.

I guess this is an existing issue though and maybe something we should fix in a separate PR.

Reply:
This is actually somewhat intentional to allow for background HTTP servers to work without breaking the sanitizer. Let's figure this out in a follow up.